### PR TITLE
LLM support: improve VGF export and calibration pipeline

### PR DESCRIPTION
**Pull request by xingguo01**
This is stacked on top of https://github.com/pytorch/executorch/pull/19029
- make non-KV-cache example inputs match the static export window
- fix PT2E calibration flow for padded prefixes
  and optional LM-Eval tasks
- update SmolLM2 export settings used by the VGF PT2E workflow
- Fix rope_theta in 135M_config.json to align with Hugging face
  model config

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Comment by zingo** [comment]
Hi @digantdesai OK to merge this?

File: examples/models/llama/eval_llama_lib.py [hunk 1]
```diff
@@ -80,7 +85,47 @@ def _model_call(self, inps):
...
+            # Reconstruct full logits by running prefixes.
+            result_logits = []
+            for pos in range(min(seq_len, self._max_seq_length)):
+                prefix = pad_to_max_len(inps[:, : pos + 1])
+                logits = self._model(prefix)
+                if logits.dim() == 3:
+                    logits = logits[:, -1, :]
+                result_logits.append(logits)
+
+            return torch.stack(result_logits, dim=1)
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply examples/models/llama/eval_llama_lib.py]: Good point. The prefix reconstruction path should be numerically equivalent for a causal model, but it is not a good default for lm-eval because it scales as one forward per prefix position.

I found one wiring issue here: the eval entrypoint forces generate_full_logits=True, but the quantized GraphModuleEvalWrapper was not receiving that config, so it defaulted to False and could hit this slow fallback. I fixed that by passing llm_config.debug.generate_full_logits into GraphModuleEvalWrapper.

Longer term, I agree a guard or warning would be useful if someone explicitly evaluates a non-KV-cache export with generate_full_logits=False at larger seq lengths.

File: examples/models/llama/model.py [hunk 2]
```diff
@@ -285,11 +286,11 @@ def get_example_inputs(self):
...
             return self.get_example_inputs_kvcache_sdpa()
         else:
-            return (
-                torch.tensor(
-                    [[1, 2, 3]], dtype=torch.long
-                ),  # tokens, with kv cache our input token length is always just 1 token.
-            )
+            max_len = getattr(self.llm_config.export, "max_seq_length", 3)
+            max_len = max(3, int(max_len))
+            example_tokens = torch.arange(max_len, dtype=torch.int32).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply examples/models/llama/model.py]: This has been addressed in the latest patchset. The non-KV-cache example input no longer unconditionally uses int32: it now uses torch.long by default, while preserving torch.int32 only for the Arm backend paths that require it (ethosu/tosa/vgf).

File: extension/llm/export/builder.py [hunk 3]
```diff
@@ -307,26 +323,31 @@ def calibrate_template(
...
 
-        for task, res in eval_results["results"].items():
-            print(f"{task}: {res}")
-        logging.info("Calibration finish...")
+            with torch.no_grad():
+                eval_results = simple_evaluate(
+                    model=eval_wrapper,
+                    tasks=calibration_tasks,
+                    limit=calibration_limit,
+                )
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply extension/llm/export/builder.py:371]: This is addressed in the latest patchset by adding a warning before calibration starts when calibration_tasks is provided without calibration_limit. I intentionally left the behavior unchanged instead of forcing a default, because existing callers may rely on lm-eval's full-dataset behavior.

File: examples/models/llama/model.py [hunk 4]
```diff
@@ -285,11 +286,21 @@ def get_example_inputs(self):
...
+            token_dtype = (
+                torch.int32
+                if (
+                    backend.ethosu.enabled
+                    or backend.tosa.enabled
+                    or backend.vgf.enabled
+                )
+                else torch.long
             )
+            example_tokens = torch.arange(max_len, dtype=token_dtype).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply examples/models/llama/model.py]: Addressed. The non-KV example input now uses max_seq_length - 1 only when enable_dynamic_shape=True, matching the existing dynamic-shape upper bound in LLMEdgeManager. Static exports still use the full max_seq_length window.

File: extension/llm/export/builder.py [hunk 5]
```diff
@@ -360,9 +387,7 @@ def pt2e_quantize(self, quantizers: Optional[List[Quantizer]]) -> "LLMEdgeManage
                 )
                 # Calibrate
                 if (
-                    self.calibration_tasks is not None
-                    and self.calibration_limit is not None
-                    and self.calibration_seq_length is not None
+                    self.calibration_seq_length is not None
                     and self.calibration_data is not None
                     and self.tokenizer_path is not None
                 ):
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply extension/llm/export/builder.py:415]: Addressed. The lm-eval-specific imports are now guarded by `if calibration_tasks:`, so prompt-only calibration can run without lm-eval installed. The broadened `pt2e_quantize()` condition now only requires `calibration_seq_length`, `calibration_data`, and `tokenizer_path` for template calibration.

File: extension/llm/export/builder.py [hunk 6]
```diff
@@ -282,23 +282,39 @@ def calibrate_template(
...
                     pos += 1
                     if pos >= len(token_list):
                         if self.generate_full_logits:
-                            token_list.append(
-                                torch.argmax(logits[:, -1], dim=-1).item()
-                            )
+                            next_token = torch.argmax(logits[:, -1], dim=-1).item()
                         else:
-                            token_list.append(torch.argmax(logits[:], dim=-1).item())
+                            next_token = torch.argmax(logits[:], dim=-1).item()
```
[the remark this answers is not in the document]
Comment (xingguo01) [reply extension/llm/export/builder.py:331]: Addressed. Static non-KV full-logits calibration now tracks `logits_token_pos` before padding and samples `logits[:, logits_token_pos]`, i.e. the last non-pad prefix token rather than the padded tail. Static non-KV calibration without full logits is rejected because that output shape cannot select the non-pad token row safely.

File: examples/models/llama/eval_llama_lib.py [hunk 7]
```diff
@@ -80,7 +85,47 @@ def _model_call(self, inps):
...
+
+            seq_len = inps.shape[-1]
+
+            def pad_to_max_len(tokens: torch.Tensor) -> torch.Tensor:
+                if self._enable_dynamic_shape:
+                    return tokens
+                token_len = tokens.shape[-1]
+                if token_len < self._max_seq_length:
+                    pad_len = self._max_seq_length - token_len
+                    pad_token = getattr(
```
Comment (digantdesai) [inline examples/models/llama/eval_llama_lib.py]: curious what's the impact on eval from padding with eos for this non-KV path with generate_full_logits=False or this is mainly for shape compatibility?

File: examples/models/llama/model.py [hunk 8]
```diff
@@ -285,11 +286,22 @@ def get_example_inputs(self):
         if self.use_kv_cache:
             return self.get_example_inputs_kvcache_sdpa()
         else:
-            return (
-                torch.tensor(
-                    [[1, 2, 3]], dtype=torch.long
-                ),  # tokens, with kv cache our input token length is always just 1 token.
+            max_seq_len = getattr(self.llm_config.export, "max_seq_length", 3)
+            max_seq_len = max(3, int(max_seq_len))
```
Comment (digantdesai) [inline examples/models/llama/model.py:291]: what's special about 3, curious.

**Comment by digantdesai** [comment]
@claude Review this PR.

**Comment by digantdesai** [comment]
Let me finish the review later today, LGTM at a high level.

File: examples/models/llama/eval_llama_lib.py (hunk 7 above)
Comment (xingguo01) [reply examples/models/llama/eval_llama_lib.py]: So yes, with static non-KV + generate_full_logits=False, EOS padding can affect the returned last-position logits if the graph uses the padded tail as the last valid token. That mode is mainly a shape-compatibility fallback, not the preferred evaluation path. For accurate static non-KV eval/calibration we should use generate_full_logits=True or pass an explicit last_valid_token_pos through the graph.

File: examples/models/llama/model.py (hunk 8 above)
Comment (xingguo01) [reply examples/models/llama/model.py:291]: Nothing special semantically. `3` preserves the old example-input length/default (`[[1, 2, 3]]`) for the generic non-KV path when no export max_seq_length is configured.

The `max(3, ...)` guard also avoids producing a too-small dynamic-shape example, because the dynamic non-KV path uses `max_seq_length - 1` as the example length. So with the minimum of 3, the dynamic example is still length 2.

For real exports, `export.max_seq_length` should be set explicitly, so this fallback only affects default/example behavior.

**Comment by xingguo01** [comment]
Thanks, I have been addressing the Claude comments and will get back to it on Monday. Need internal reviews before upstreaming.

**Comment by xingguo01** [comment]
@claude here is the reply, implemented in the new commits: Thanks for the detailed review.

I have addressed the actionable items in the latest update:

1. Static non-KV eval with `generate_full_logits=False`
   `GraphModuleEvalWrapper` now rejects this combination, matching the builder-side calibration guard. Static non-KV eval requires `generate_full_logits=True` so the wrapper can use the full `[batch, seq, vocab]` output and avoid reading logits from the padded tail.

2. Example token IDs and `max(3, ...)`
   The example token generation is now bounded by the model vocab size instead of `% 100 + 1`, avoiding out-of-range token IDs for small-vocab test configs. I also added a short inline comment explaining that `3` preserves the historical minimum example input length.

3. Duplicate calibration logging
   Removed the unconditional duplicate log. The calibration config is now logged only when calibration actually runs.

4. Calibration prefix validation
   Moved `_check_calibration_prefix_options()` to the start of `pt2e_calibrate()` so unsupported static non-KV/non-full-logits calibration fails before tokenization or other work.

5. Device behavior
   The calibration path intentionally passes `device="cpu"` for stability because some exported graphs can contain CPU-defaulting ops. The general eval wrapper still preserves the existing default behavior when no device is passed, so existing eval callers keep their current device selection behavior. I can add this note to the PR description.

6. Full-logits static padding
   Agreed. For `generate_full_logits=True`, padding is shape compatibility only. lm-eval scores the real token positions, and causal masking means trailing pad/EOS tokens do not affect earlier logits.

I left the eval-time and calibration-time padding helpers separate for now because the contexts differ slightly.

**Comment by xingguo01** [comment]
@digantdesai all comments addressed, highly appreciated if you let me know if there are additional comments on this PR.

**Review by digantdesai — APPROVED** [review]

**Comment by zingo** [comment]
For the record, this broke  

[trunk / test-arm-backend-ethos-u (test_smaller_stories_llama) / linux-job](https://hud.pytorch.org/pr/pytorch/executorch/19157#77888387512) ([gh](https://github.com/pytorch/executorch/actions/runs/26455557146/job/77888387512))
##[error]The operation was canceled.

See https://hud.pytorch.org/hud/pytorch/executorch/main/1?per_page=50&name_filter=test-arm%7Ccortex-m%7Csize&useRegexFilter=true

Work/fix is ongoing https://github.com/pytorch/executorch/pull/20026 Hopefully that will fix it again.
So in case of cherry-picks or other things in the future you probably need both PRs
